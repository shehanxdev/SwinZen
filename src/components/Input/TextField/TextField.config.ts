//TODO::add mappings and size configurations in here, when the design system grows further.
