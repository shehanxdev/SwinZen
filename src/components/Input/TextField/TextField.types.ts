import React, { ComponentProps } from 'react';
import { TextInput } from 'react-native-paper';
import { StyleProp, ViewStyle } from 'react-native/types';

import { Color } from '@sz/constants';

type RNTextInputProps = ComponentProps<typeof TextInput>;

type WithRNTextInputProps = Pick<
  RNTextInputProps,
  | 'defaultValue'
  | 'value'
  | 'disabled'
  | 'placeholder'
  | 'onChangeText'
  | 'onBlur'
  | 'onFocus'
  | 'secureTextEntry'
  | 'maxLength'
  | 'keyboardType'
  | 'onSubmitEditing'
  | 'onEndEditing'
  | 'blurOnSubmit'
  | 'autoCapitalize'
  | 'textContentType'
  | 'autoComplete'
  | 'testID'
  | 'autoCorrect'
  | 'numberOfLines'
  | 'multiline'
  | 'returnKeyType'
  | 'returnKeyLabel'
>;

export interface TextFieldProps extends WithRNTextInputProps {
  label?: string | React.ReactNode;
  innerTextInputStyles?: StyleProp<ViewStyle>;
  helperText?: string;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  textColor?: Color;
  disabledColor?: Color;
  labelColor?: Color;
  helperTextColor?: Color;
  error?: boolean;
  backgroundColor?: Color;
  outlineColor?: Color;
  activeOutlineColor?: Color;
  onRightIconPress?: () => void;
}
