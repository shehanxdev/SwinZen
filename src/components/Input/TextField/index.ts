export * from './TextField.component';
export * from './TextField.types';
