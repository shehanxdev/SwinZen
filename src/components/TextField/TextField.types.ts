import React, { ComponentProps } from 'react';
import { TextInput } from 'react-native-paper';

import { Color } from '@sz/constants';

type RNTextInputProps = ComponentProps<typeof TextInput>;

type WithRNTextInputProps = Pick<
  RNTextInputProps,
  | 'defaultValue'
  | 'value'
  | 'disabled'
  | 'label'
  | 'placeholder'
  | 'onChangeText'
  | 'onBlur'
  | 'onFocus'
  | 'secureTextEntry'
  | 'maxLength'
  | 'label'
  | 'keyboardType'
  | 'onSubmitEditing'
  | 'onEndEditing'
  | 'blurOnSubmit'
  | 'autoCapitalize'
  | 'textContentType'
  | 'autoComplete'
  | 'testID'
  | 'autoCorrect'
  | 'numberOfLines'
  | 'multiline'
>;

export interface TextFieldProps extends WithRNTextInputProps {
  helperText?: string;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  textColor?: Color;
  disabledColor?: Color;
  labelColor?: Color;
  helperTextColor?: Color;
  error?: boolean;
  backgroundColor?: Color;
  onRightIconPress?: () => void;
}
