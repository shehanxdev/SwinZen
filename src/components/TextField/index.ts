export * from './TextField.component';
export * from './TextField.types';
export * from './TextArea.component';
