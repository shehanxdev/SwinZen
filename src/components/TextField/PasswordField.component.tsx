import React, { forwardRef, useState } from 'react';
import { TextInput as RNTextInput } from 'react-native';

import { Color } from '@sz/constants';

import { PasswordHideIcon, PasswordRevealIcon, SvgIconProps } from '../Icon';
import { TextField } from './TextField.component';
import { TextFieldProps } from './TextField.types';

interface PasswordFieldProps extends TextFieldProps {
  passwordIconColor?: Color;
}

export const PasswordField = forwardRef<RNTextInput, PasswordFieldProps>(function PasswordInput(
  { passwordIconColor = Color.Neutral.Sz500, ...props },
  ref,
) {
  const [enablePasswordIcon, setEnablePasswordIcon] = useState(true);

  const iconProps: SvgIconProps = {
    color: passwordIconColor,
  };

  return (
    <TextField
      ref={ref}
      onRightIconPress={() => setEnablePasswordIcon(!enablePasswordIcon)}
      rightIcon={enablePasswordIcon ? <PasswordRevealIcon {...iconProps} /> : <PasswordHideIcon {...iconProps} />}
      secureTextEntry={enablePasswordIcon}
      {...props}
    />
  );
});
