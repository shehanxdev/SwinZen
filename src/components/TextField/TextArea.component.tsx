import React, { forwardRef } from 'react';
import { TextInput as RNTextInput } from 'react-native';

import { TextField } from './TextField.component';
import { TextFieldProps } from './TextField.types';

export const TextArea = forwardRef<RNTextInput, TextFieldProps>(function TextAreainput(props, ref) {
  return <TextField ref={ref} multiline={true} innerTextInputStyle="h-[132px] p-4 m-0" {...props} />;
});
