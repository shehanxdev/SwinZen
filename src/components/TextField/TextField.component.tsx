import React, { cloneElement, forwardRef, useMemo } from 'react';
import { TextInput as RNTextInput, View } from 'react-native';
import { TextInput } from 'react-native-paper';

import { tw } from '@sz/config';
import { Color, TextAlignment, TextVariant } from '@sz/constants';

import { Text } from '../Typography';
import { TextFieldProps } from './TextField.types';

//TODO::handle scroll on focus
export const TextField = forwardRef<RNTextInput, TextFieldProps>(function AppTextField(
  {
    backgroundColor = Color.Neutral.Sz1000,
    defaultValue,
    value,
    disabled = false,
    label,
    helperText,
    helperTextColor = Color.Neutral.Sz200,
    leftIcon,
    labelColor = Color.Neutral.Sz200,
    onChangeText,
    onBlur,
    onFocus,
    rightIcon = null,
    secureTextEntry,
    testID,
    textColor = Color.Neutral.Sz500,
    disabledColor = Color.Neutral.Sz700,
    blurOnSubmit,
    onSubmitEditing,
    activeOutlineColor = Color.Neutral.Sz600,
    outlineColor = Color.Transparency.full,
    onEndEditing,
    maxLength,
    keyboardType,
    error,
    placeholder,
    autoCapitalize,
    textContentType,
    autoComplete,
    onRightIconPress,
    autoCorrect,
    multiline,
    numberOfLines,
    returnKeyType,
    returnKeyLabel,
  }: TextFieldProps,
  ref,
) {
  const labelTextComponent = useMemo(
    () => (
      <Text variant={TextVariant.Body2SemiBold} color={labelColor} textAlign={TextAlignment.Auto}>
        {label}
      </Text>
    ),
    [labelColor, label, error],
  );

  const helperTextComponent = useMemo(
    () => (
      <Text
        variant={TextVariant.Labels}
        color={error ? Color.Error.SzMain : helperTextColor}
        textAlign={TextAlignment.Auto}>
        {helperText}
      </Text>
    ),
    [helperText, helperTextColor, error],
  );

  /* Design system updates the color of the left icon when the text input gets disabled. So rather than handling the color change within the place TextField being used,
   * create a new React element using original element using React.cloneElement using below two functions.
   */
  const leftIconComponent = useMemo(() => {
    return (
      leftIcon &&
      cloneElement(leftIcon as any, {
        ...(disabled && { color: disabledColor }),
      })
    );
  }, [leftIcon, disabled]);

  const rightIconComponent = useMemo(() => {
    return (
      rightIcon &&
      cloneElement(rightIcon as any, {
        ...(disabled && { color: disabledColor }),
      })
    );
  }, [rightIcon, disabled]);

  return (
    <View>
      {labelTextComponent}
      <TextInput
        style={tw`h-12.5 p-0 m-0`}
        ref={ref}
        mode="outlined"
        secureTextEntry={secureTextEntry}
        defaultValue={defaultValue}
        textContentType={textContentType}
        disabled={disabled}
        autoComplete={autoComplete}
        value={value}
        onChangeText={onChangeText}
        onBlur={onBlur}
        onFocus={onFocus}
        placeholder={placeholder}
        theme={{
          colors: {
            primary: Color.Primary.Sz800,
            placeholder: Color.Neutral.Sz500, //Figma does not specify this value. Update accordinly
            text: disabled ? Color.Neutral.Sz700 : textColor,
            background: backgroundColor,
            disabled: Color.Neutral.Sz700,
          },
          roundness: 10,
          fonts: {
            regular: tw`text-[16px] font-normal p-0 leading-[18px]`,
          },
        }}
        autoCapitalize={autoCapitalize}
        onEndEditing={onEndEditing}
        onSubmitEditing={onSubmitEditing}
        maxLength={maxLength}
        keyboardType={keyboardType}
        blurOnSubmit={blurOnSubmit}
        testID={testID}
        outlineColor={error ? Color.Error.SzMain : outlineColor}
        activeOutlineColor={error ? Color.Error.SzMain : activeOutlineColor}
        autoCorrect={autoCorrect}
        multiline={multiline}
        numberOfLines={numberOfLines}
        left={leftIcon && <TextInput.Icon name={() => leftIconComponent} style={{ marginTop: '50%' }} />}
        right={
          rightIcon && (
            <TextInput.Icon name={() => rightIconComponent} onPress={onRightIconPress} style={{ marginTop: '50%' }} />
          )
        }
        returnKeyType={returnKeyType}
        returnKeyLabel={returnKeyLabel}
      />
      {helperTextComponent}
    </View>
  );
});
