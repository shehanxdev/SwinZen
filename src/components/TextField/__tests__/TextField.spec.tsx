import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';

import { Color } from '@sz/constants';

import { TextFieldProps } from '../TextField.types';
import { TextField } from './../TextField.component';

describe('TextField Component', () => {
  let value = '';
  const testID = 'TEST_ID';
  const mockOnChangeFunction = jest.fn();

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const otherProps: TextFieldProps = {
    defaultValue: 'Default value',
    helperText: 'Helper text',
    label: 'Label',
    onChangeText: v => {
      mockOnChangeFunction();
      value = v;
    },
    textColor: Color.Neutral.Sz200,
    disabledColor: Color.Neutral.Sz100,
    labelColor: Color.Tertiary.Sz200,
    helperTextColor: Color.Tertiary.Sz300,
  };

  const getRenderedTextFieldComponent = (props?: Partial<TextFieldProps>) =>
    render(
      <TextField
        {...props}
        testID={testID}
        defaultValue={otherProps.defaultValue}
        helperText={otherProps.helperText}
        onChangeText={otherProps.onChangeText}
      />,
    );

  test(`should render correctly`, () => {
    const rendered = getRenderedTextFieldComponent();
    const renderedTree = rendered.toJSON();
    expect(renderedTree).toMatchSnapshot();
  });

  test('should allow to change text in the TextInput', () => {
    const { getByTestId } = getRenderedTextFieldComponent();
    const foundTextField = getByTestId(testID);

    const text = (Math.random() + 1).toString(36).substring(7);

    fireEvent.changeText(foundTextField, text);
    expect(foundTextField.props.value).toBe(text);
  });

  test('should allow text to be deleted', () => {
    const { getByTestId } = getRenderedTextFieldComponent();
    const foundTextField = getByTestId(testID);

    const text = (Math.random() + 1).toString(36).substring(7);

    fireEvent.changeText(foundTextField, text);
    expect(foundTextField.props.value).toBe(text);

    fireEvent.changeText(foundTextField, '');
    expect(foundTextField.props.value).toBe('');
  });

  test('should execute onChangeText when user change text', () => {
    const { getByTestId } = getRenderedTextFieldComponent();
    const foundTextField = getByTestId(testID);

    const text = (Math.random() + 1).toString(36).substring(7);

    fireEvent.changeText(foundTextField, text);
    expect(mockOnChangeFunction).toHaveBeenCalled();
    expect(foundTextField.props.value).toBe(value);
  });
});
