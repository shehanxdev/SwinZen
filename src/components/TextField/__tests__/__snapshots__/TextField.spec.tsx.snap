// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Component should render correctly 1`] = `
<View>
  <Text
    style={
      [
        {
          "color": "#E9ECEF",
          "lineHeight": 24,
        },
        {
          "fontFamily": "SourceSansPro-SemiBold",
          "fontSize": 16,
          "fontWeight": "600",
          "textAlign": "auto",
        },
      ]
    }
  />
  <View
    style={
      {
        "marginBottom": 0,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 6,
          },
          {
            "backgroundColor": "#16501D",
            "borderColor": "#ffffff00",
            "borderRadius": 10,
            "borderWidth": 1,
          },
        ]
      }
      testID="text-input-outline"
    />
    <View>
      <View
        style={
          [
            {
              "paddingBottom": 0,
            },
            {
              "minHeight": 48,
              "paddingTop": 8,
            },
          ]
        }
      >
        <TextInput
          defaultValue="Default value"
          editable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholderTextColor="#ADB5BD"
          selectionColor="#6C757D"
          style={
            [
              {
                "flexGrow": 1,
                "margin": 0,
                "paddingHorizontal": 14,
                "zIndex": 1,
              },
              {
                "height": 48,
              },
              {
                "paddingBottom": 16,
                "paddingTop": 16,
              },
              {
                "color": "#DEE2E6",
                "fontFamily": "System",
                "fontSize": 16,
                "fontVariant": undefined,
                "fontWeight": undefined,
                "lineHeight": 18,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
            ]
          }
          testID="TEST_ID"
          underlineColorAndroid="transparent"
          value="Default value"
        />
      </View>
    </View>
  </View>
  <Text
    style={
      [
        {
          "color": "#E9ECEF",
          "lineHeight": 24,
        },
        {
          "fontFamily": "SourceSansPro-Regular",
          "fontSize": 15,
          "fontWeight": "600",
          "textAlign": "auto",
        },
      ]
    }
  >
    Helper text
  </Text>
</View>
`;
